ref: `exp run --rev` change (#2383)
iterative/dvc.org: content/docs/command-reference/exp/run.md

You can now use `dvc exp run` to begin the experiment. If the process gets
interrupted (e.g. with `[Ctrl] C` or by an error), all the checkpoints so far
will be preserved. When a run finishes normally, a final checkpoint will be
added (if needed) to wrap up the experiment. Subsequent uses of `dvc exp run`
will resume from this point (using the latest cached versions of all outputs).

List previous checkpoints with `dvc exp show`. To continue from a previous
checkpoint, you must first `dvc exp apply` it before using `dvc exp run`. For

> `--queue` or `--temp` runs (see below), use `--rev` instead to specify the
> checkpoint to continue from.

Reviewer: ```suggestion
`--queue` or `--temp` runs (see next section), use `--rev` instead to specify
the checkpoint to continue from.
```

Reviewer: I guess this is the last bit. Will commit and merge 🚀

You can now use `dvc exp run` to begin the experiment. If the process gets
interrupted (e.g. with `[Ctrl] C` or by an error), all the checkpoints so far
will be preserved. When a run finishes normally, a final checkpoint will be
added (if needed) to wrap up the experiment. Subsequent uses of `dvc exp run`
will resume from this point (using the latest cached versions of all outputs).

> List previous checkpoints with `dvc exp show`. To continue from a previous
> checkpoint, you must first `dvc exp apply` it before using `dvc exp run`. For

Reviewer: The first sentence still feels a bit disconnected. How about something like
```suggestion
List previous checkpoints with `dvc exp show`. To continue from any of
them, you must first `dvc exp apply` it before using `dvc exp run`.
```

> List previous checkpoints with `dvc exp show`. To continue from a previous
> checkpoint, you must first `dvc exp apply` it before using `dvc exp run`. For

Reviewer: The first sentence still feels a bit disconnected. How about something like
```suggestion
List previous checkpoints with `dvc exp show`. To continue from any of
them, you must first `dvc exp apply` it before using `dvc exp run`.
```

Reviewer 2: "them -> it" sounds awkward - I'd prefer slight repetition instead.

Reviewer: True 👍

> checkpoint, you must first `dvc exp apply` it before using `dvc exp run`. For
> convenience, when running experiments outside of your workspace via
> `--queue/--run-all` or `--temp`, you can use `--rev` instead to specify the
> checkpoint from which you want to continue.

Reviewer: I think we can simplify this part a little, for example:
```suggestion
checkpoint, you must first `dvc exp apply` it before using `dvc exp run`. For
`--queue` or `--temp` runs (see following section), use `--rev` instead to
specify the checkpoint from which you want to continue.
```

> checkpoint, you must first `dvc exp apply` it before using `dvc exp run`. For
> convenience, when running experiments outside of your workspace via
> `--queue/--run-all` or `--temp`, you can use `--rev` instead to specify the
> checkpoint from which you want to continue.

Reviewer: I think we can simplify this part a little, for example:
```suggestion
checkpoint, you must first `dvc exp apply` it before using `dvc exp run`. For
`--queue` or `--temp` runs (see following section), use `--rev` instead to
specify the checkpoint from which you want to continue.
```

Reviewer 2: even shorter:

```suggestion
checkpoint, you must first `dvc exp apply` it before using `dvc exp run`. For
`--queue` or `--temp` runs (see below), use `--rev` instead to specify the
checkpoint to continue from.
```

Reviewer: 👍 except that it's not clear what "see below" refers to. Will re-suggest ⌛

> 
> ⚠️ Note that `--rev` is disallowed when running in your workspace, since using
> `--rev` in this way could potentially lead to discarding and overwriting changes
> in your workspace.

Reviewer 2: ```suggestion
```

Alternatively, use `--reset` to start over (discards previous checkpoints and

> their outputs, leaving things as they were before the first `dvc exp run`). This
> is useful for re-training ML models, for example.

Reviewer: ```suggestion
their outputs,). This is useful for re-training ML models, for example.
```

Reviewer 2: ```suggestion
their outputs). This is useful for re-training ML models, for example.
```

Reviewer: Thanks for typo fix!

You can now use `dvc exp run` to begin the experiment. If the process gets
interrupted (e.g. with `[Ctrl] C` or by an error), all the checkpoints so far
will be preserved. When a run finishes normally, a final checkpoint will be

> added (if needed) to wrap up the experiment. Following uses of `dvc exp run`
> will continue from this point (using the latest cached versions of all outputs).

Reviewer 2: Had to re-read this several times to understand the original wording. How about these synonyms which can't be misinterpreted?

```suggestion
added (if needed) to wrap up the experiment. Subsequent uses of `dvc exp run`
will resume from this point (using the latest cached versions of all outputs).
```

> To continue from a previous checkpoint (list them with `dvc exp show`), when
> running experiments in your workspace, you must first use `dvc exp apply` to
> apply the checkpoint from which you want to resume before using `dvc exp run`.
> When running experiments outside of your workspace via `--queue/--run-all` or
> `--temp`, you can use `--rev` to directly specify the checkpoint from which you
> want to resume.

Reviewer 2: ```suggestion
List previous checkpoints with `dvc exp show`. To continue from a previous
checkpoint when running experiments in your workspace, you must first use
`dvc exp apply` to apply the checkpoint from which you want to resume before
using `dvc exp run`. When running experiments outside of your workspace via
`--queue/--run-all` or `--temp`, you can use `--rev` to directly specify the
checkpoint from which you want to resume.
```

Reviewer: Improved for sure. But this section is about checkpoints, not about `exp show`. Shouldn't even be in `exp run` probably (WIP in #2373). Maybe the `exp show` brief mention should go earlier in some other paragraph?

Reviewer: Let's remove ~~when running experiments in your workspace~~ ? The queue/temp section comes after this one so I doubt people will understand what it means. Maybe a paragraph like this instead "(except in `--queue/temp` runs, see next section)"

Reviewer: "use `exp apply` to apply" sounds redundant I think.

Reviewer: How about just "you must `exp apply` it first" ?

Reviewer: Cc @casperdcl those ☝️ were meant as ideas for you to update your suggestion 🙂

Reviewer: Also:
- ~~directly~~
- "from which you want to ~~resume~~ continue" (consistency)

Reviewer 2: ```suggestion
List previous checkpoints with `dvc exp show`. To continue from a previous
checkpoint, you must first `dvc exp apply` it before using `dvc exp run`. For
convenience, when running experiments outside of your workspace via
`--queue/--run-all` or `--temp`, you can use `--rev` instead to specify the
checkpoint from which you want to continue.
```

This is a command reference so I think it's fine to talk about `--queue/--run-all/--temp` without adding "(see below)".

Reviewer: Nice, committing. BTW I'm just trying to give some recs these first times, but you're the reviewer  @casperdcl, so up to you 🙂 (unless someone detects errors or has strong opinions, in which case lets discuss as needed). Anyway, committing this one for now.

Reviewer: Nice, committing. BTW I'm just trying to give some recs these first times, but you're the reviewer  @casperdcl, so up to you 🙂 (unless someone detects errors or has strong opinions, in which case lets discuss as needed).

> ⚠️ Note that `--rev` is disallowed when running in your workspace, since using
> `--rev` in this way could potentially lead to discarding and overwriting changes
> in your workspace.

Reviewer 2: ```suggestion
⚠️ Note that `--rev` is disallowed when running in your workspace in order to
prevent uncommitted changes potentially being discarded/overwritten.
```

is "uncommitted" an acceptable term to use according to DVC nomenclature @jorgeorpinel?

Reviewer: > is "uncommitted" an acceptable term

I'd just say "changes". We want to simplify the language as much as possible.

But TBH I'm not sure we need this whole note (I know @dberenbaum requested some motivation but also it looks a bit like an implementation detail, and we avoid those in docs in general).

Reviewer 2: Presumably if you try to use `--rev` incorrectly you'll get a warning message saying "please use `exp apply` instead"?

Reviewer 2: AFAIK actually https://github.com/iterative/dvc/blob/54c0675fc664cd97394a1f789aaa385a5205a304/dvc/command/experiments.py#L517-L521 is this a bug? It only checks for `queue or tmp_dir` but not `run_all`

Reviewer 3: Yes, I agree this whole note seems unnecessary (the paragraph above already specifies that `--rev` is for use outside the workspace).

@casperdcl I think `--run-all` will silently ignore `--rev`. It actually probably silently ignores lots of arguments actually now that I think about since it's kind of a unique flag that doesn't create an experiment but runs existing ones. Not sure what's best here, although the `--queue` and `--run-all` flags might ultimately get replaced by a `exp queue` command.

Reviewer: > Should we just remove this note/paragraph?
agree this whole note seems unnecessary

☠️ it!

> actually https://github.com/iterative/dvc/blob/54c0675fc664cd97394a1f789aaa385a5205a304/dvc/command/experiments.py#L517-L521 is this a bug?

@casperdcl did you have a chance to try it locally? Always best to use the tool and double check during the docs process, I think. But if still needed, please tag **pmrowla** for implementation Qs.

Reviewer: > --run-all will silently ignore --rev ... probably silently ignores lots of arguments...
it's kind of a unique flag that doesn't create an experiment but runs existing ones

@dberenbaum great Qs but I think that's a slightly different Q. Let me reopen that as a separate review...

Reviewer: > Let me reopen that as a separate review

Actually @casperdcl already mentioned it separately in https://github.com/iterative/dvc.org/pull/2383#pullrequestreview-644051647. I couldn't find a good place in the code for that review anyway, seems like it's not in the scope of this PR? But

> the --queue and --run-all flags might ultimately get replaced by a exp queue command

Is there an issue to track that @dberenbaum ? Thanks 🙂

Author: https://github.com/iterative/dvc/issues/5615

Reviewer 2: nuked.

> Alternatively, use `--reset` to start over (discards previous checkpoints and
> deletes `checkpoint` outputs, like the first `dvc exp run`) — useful for
> re-training ML models, for example.

Reviewer 2: ```suggestion
Alternatively, use `--reset` to start over (discards previous checkpoints and
their outputs, like things as they were before the first `dvc exp run`). This is
useful for re-training ML models, for example.
```

Reviewer: Def. better here as well, but

> like things as they were before

We literally meant "like the first `dvc exp run` (does)". Meaning that the first `dvc exp run` deletes checkpoint outputs. The above suggestion would say "discards things as they were before". Maybe let's try to simplify the sentence, maybe remove split into 2 without parentheses? Not sure

Reviewer 2: I meant to write "leaving," not "like." Better now?

Reviewer: Yes, thanks. Let me commit this but actually I realized we decided at some point not to document the "first `exp run` edge case so we could just remove that parenthesis 😆. I'll make a separate suggestion... ⌛

  is processed serially).

- `-r <commit>`, `--rev <commit>` - continue an experiment from a specific

>   checkpoint name or hash (`commit`) in `--queue` or `--temp` runs.

Reviewer: Or maybe

continue a checkpoint experiment run with  `--queue` or `--temp` from a specific checkpoint name or hash (`commit`).

Reviewer 2: Also `--run-all`?

Reviewer: The type of experiment/run is "queued experiment" so `--queue` should be clear enough, I think. Sounds like `--run-all` may get extracted to a separate subcommand in the near future anyway.

  is processed serially).

- `-r <commit>`, `--rev <commit>` - continue an experiment from a specific

>   checkpoint name or hash (`commit`). This is required to resume experiments
>   from `--queue` or `--temp` runs. `--rev` is explicitly disallowed when running
>   experiments in your workspace (`dvc exp apply` should be used instead).

Reviewer: ```suggestion
  checkpoint name or hash (`commit`) in `--queue` or `--temp` runs.
```
